In [2]:
# https://www.kaggle.com/datasets/08163f86754da93458b7dfb3e1c80258069edd817a00db655f3ccc3da37c87b5?resource=download

# Download Maths.csv dataset from this link and put it into notebook directory.

In [35]:
import pandas as pd

RANDOM_STATE = 42
maths = pd.read_csv('Maths.csv')
maths.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [39]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

CURRENT = "G3"
categorical_columns = ["school", "sex", "address", "famsize", "Pstatus", "Mjob", "Fjob", "reason", "guardian", 
                       "schoolsup", "famsup", "paid", "activities", "nursery", "higher", "internet", "romantic"]
X = maths.copy()
X = pd.get_dummies(X, columns = categorical_columns)
y = X[CURRENT]
X = X.drop(CURRENT, axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=RANDOM_STATE)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)                      
X_test = scaler.transform(X_test)                      

In [46]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(random_state=RANDOM_STATE).fit(X_train, y_train)
test_score = clf.score(X_test, y_test)
train_score = clf.score(X_train, y_train)
name = type(clf).__name__
print(f"Accuracy of {name} model:\nTrain set: {train_score}\nTest set: {test_score}")

Accuracy of DecisionTreeClassifier model:
Train set: 1.0
Test set: 0.26582278481012656


In [48]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(random_state=RANDOM_STATE).fit(X_train, y_train)
test_score = clf.score(X_test, y_test)
train_score = clf.score(X_train, y_train)
name = type(clf).__name__
print(f"Accuracy of {name} model:\nTrain set: {train_score}\nTest set: {test_score}")

Accuracy of LogisticRegression model:
Train set: 0.7911392405063291
Test set: 0.22784810126582278


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [49]:
from sklearn.ensemble import AdaBoostClassifier

clf = AdaBoostClassifier(random_state=RANDOM_STATE,n_estimators=100).fit(X_train, y_train)
test_score = clf.score(X_test, y_test)
train_score = clf.score(X_train, y_train)
name = type(clf).__name__
print(f"Accuracy of {name} model:\nTrain set: {train_score}\nTest set: {test_score}")

Accuracy of AdaBoostClassifier model:
Train set: 0.2626582278481013
Test set: 0.17721518987341772


# Results
We used `DecisionTreeClassifier`, `AdaBoostClassifier` and `LogisticRegression`.

Boosting algorithm had the worst performance with training accuracy of 26% and testing accuracy of 17%. 
Logistic regression model had 79% accuracy on training set and 22.7% on testing set.
DecisionTreeClassifier yielded the best results with 100% accuracy on training set and 26.5% accuracy on testing set.
